ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Crypto.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PatternSynonyms       #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+-- | Tests consensus-specific crypto operations in relationship with blocks/headers.
+module Test.Consensus.Cardano.Crypto (tests) where
+
+import           Cardano.Crypto.VRF (sizeCertVRF)
+import           Cardano.Crypto.VRF.Praos (certSizeVRF)
+import           Data.Function ((&))
+import           Ouroboros.Consensus.Cardano.Block (CardanoHeader,
+                     StandardCrypto, pattern HeaderAllegra,
+                     pattern HeaderAlonzo, pattern HeaderBabbage,
+                     pattern HeaderByron, pattern HeaderConway,
+                     pattern HeaderMary, pattern HeaderShelley)
+import           Ouroboros.Consensus.Shelley.Ledger.Block (Header (..))
+import           Ouroboros.Consensus.Shelley.Protocol.Abstract
+                     (pTieBreakVRFValue)
+import           Ouroboros.Consensus.Shelley.Protocol.Praos ()
+import           Test.Consensus.Cardano.Generators ()
+import           Test.QuickCheck (Property, label, property, (===))
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+tests :: TestTree
+tests =
+    testGroup "Cardano Crypto" [
+          testProperty "era-dependent VRF" prop_VRFCryptoDependsOnBlockEra
+    ]
+
+-- | Check Babbage and Conway blocks use different VRF crypto.
+--
+-- 1. generate (forge?) babbage or conway headers
+--   - those should contain different VRF proofs (because they are supposed to use different algorithms)
+--   - the VRF proof should be invalid for the other era
+-- 2. call some header validation functions that's calling VRF certificate check
+-- 3. assert that different VRF function is called for each era
+--    - the header should be valid
+--
+-- * why not mock everything? because it does not check that we are implementing things correctly for Conway
+--   we would like to test the dispatchign induced by this type, to make clear Conway relies on different
+--   crypto primitives
+--
+-- What needs to change is this type:
+--
+-- @@
+-- type CardanoShelleyEras c =
+--   '[ ShelleyBlock (TPraos c) (ShelleyEra c)
+--    , ShelleyBlock (TPraos c) (AllegraEra c)
+--    , ShelleyBlock (TPraos c) (MaryEra c)
+--    , ShelleyBlock (TPraos c) (AlonzoEra c)
+--    , ShelleyBlock (Praos c)  (BabbageEra c)
+--    , ShelleyBlock (Praos c)  (ConwayEra c)
```

Comment:
> Check Babbage and Conway blocks use different VRF crypto.

I don't understand why they use different crypto? To me it seems the same, both use `PraosCrypto` right?

Reply:
That's the crux of the issue: They both use `PraosCrypto` and the logic is the same, but the underlying types of VRF and KES are different so we need to find a way to have different _implementations_ for different eras.